the scheduler framework does not allow adding indexers in scheduler plugins

### What happened?

I'm writing a scheduler plugin using the scheduler framework, and in my plugin I'd like to add a [Indexer](https://pkg.go.dev/k8s.io/client-go/tools/cache#Indexer) using the `SharedInformerFactory` provided by the `framework.Handle`.
Adding such a Indexer lead to a runtime crash like
```
panic: assignment to entry in nil map

goroutine 1 [running]:
k8s.io/client-go/tools/cache.(*threadSafeMap).AddIndexers(0xc0008afb30, 0xc00099ade8?)
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/client-go/tools/cache/thread_safe_store.go:249 +0x247
k8s.io/client-go/tools/cache.(*cache).AddIndexers(0x1cd9840?, 0xc0008e98c0?)
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/client-go/tools/cache/store.go:216 +0x22
sigs.k8s.io/scheduler-plugins/pkg/noderesourcetopology.New({0x226fa48?, 0xc0005424b0?}, {0x228cee0, 0xc000810700})
	/go/src/sigs.k8s.io/scheduler-plugins/pkg/noderesourcetopology/plugin.go:110 +0x3a6
k8s.io/kubernetes/pkg/scheduler/framework/runtime.NewFramework(0xc0008afe90, 0xc00099b348, {0xc0004cac60, 0xc, 0x40e147?})
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/kubernetes/pkg/scheduler/framework/runtime/framework.go:323 +0x92b
k8s.io/kubernetes/pkg/scheduler/profile.newProfile({{0xc000056270, 0x14}, 0xc0003f7b00, {0xc0000b0600, 0x8, 0x8}}, 0x400?, 0x1?, {0xc000b07730, 0xb, ...})
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/kubernetes/pkg/scheduler/profile/profile.go:40 +0x138
k8s.io/kubernetes/pkg/scheduler/profile.NewMap({0xc0008aff20, 0x1, 0x1c83d60?}, 0xc000b073c8?, 0xc000b072a0?, {0xc00099b730, 0xb, 0xb})
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/kubernetes/pkg/scheduler/profile/profile.go:57 +0x16e
k8s.io/kubernetes/pkg/scheduler.(*Configurator).create(0xc00099b918)
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/kubernetes/pkg/scheduler/factory.go:143 +0xb93
k8s.io/kubernetes/pkg/scheduler.New({0x2291970?, 0xc00069e780}, {0x228d1d8, 0xc0001c4e10}, {0x227ef70, 0xc0001c4eb0}, 0xc0006a1b80, 0xc0004440a0?, {0xc000b07c58, 0xa, ...})
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/kubernetes/pkg/scheduler/scheduler.go:267 +0x485
k8s.io/kubernetes/cmd/kube-scheduler/app.Setup({0x2280cb8, 0xc00062b180}, 0xc000630b80, {0xc00041ee00, 0x2, 0xc0001708c0?})
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/kubernetes/cmd/kube-scheduler/app/server.go:317 +0x5fe
k8s.io/kubernetes/cmd/kube-scheduler/app.runCommand(0x6b9320?, 0xc000630b80, {0xc00041ee00, 0x2, 0x2})
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/kubernetes/cmd/kube-scheduler/app/server.go:128 +0x185
k8s.io/kubernetes/cmd/kube-scheduler/app.NewSchedulerCommand.func1(0xc000378280?, {0xc0001c31a0?, 0x2?, 0x2?})
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/k8s.io/kubernetes/cmd/kube-scheduler/app/server.go:80 +0x29
github.com/spf13/cobra.(*Command).execute(0xc000378280, {0xc0001900a0, 0x2, 0x2})
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/github.com/spf13/cobra/command.go:856 +0x67c
github.com/spf13/cobra.(*Command).ExecuteC(0xc000378280)
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/github.com/spf13/cobra/command.go:974 +0x3b4
github.com/spf13/cobra.(*Command).Execute(...)
	/go/src/sigs.k8s.io/scheduler-plugins/vendor/github.com/spf13/cobra/command.go:902
main.main()
	/go/src/sigs.k8s.io/scheduler-plugins/cmd/noderesourcetopology-plugin/main.go:51 +0x14a
```
The reason for this is that [the scheduler code initializes the cache.Indexers value with `nil`](https://github.com/kubernetes/kubernetes/blob/v1.25.0-alpha.1/pkg/scheduler/scheduler.go#L513)

### What did you expect to happen?

I'm not sure the scheduler framework wants to explicitely disallow adding Indexers using the `client-go` facilities. If this is the case, it should be more explicit somehow because the current behaviour is surprising.
If the scheduler framework doesn't want to add this limitation, then adding indexers should not lead to runtime crashes

### How can we reproduce it (as minimally and precisely as possible)?

add a new scheduler plugin trying to add indexers. Sample code (relevant minimal portions only)
```
func New(args runtime.Object, handle framework.Handle) (framework.Plugin, error) {
	klog.V(5).InfoS("Creating new custom plugin")

	podInformer := handle.SharedInformerFactory().Core().V1().Pods()
	podInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
		"nodeName": indexByPodNodeName,
	})

	podIndexer := podInformer.Informer().GetIndexer()

	// ...
}

// ...

func indexByPodNodeName(obj interface{}) ([]string, error) {
	pod, ok := obj.(*corev1.Pod)
	if !ok {
		return []string{}, nil
	}
	if len(pod.Spec.NodeName) == 0 {
		return []string{}, nil
	}
	return []string{pod.Spec.NodeName}, nil
}
```

### Anything else we need to know?

I'm not sure this is a real bug and not a deliberate decision; a `git log` research was inconcusive, and a (admittedly quick) search in the past issues was not fruitful either.

The fix could be as trivial as
```
diff --git a/pkg/scheduler/scheduler.go b/pkg/scheduler/scheduler.go
index 84e75ecf84c..a195400e117 100644
--- a/pkg/scheduler/scheduler.go
+++ b/pkg/scheduler/scheduler.go
@@ -688,5 +688,5 @@ func newPodInformer(cs clientset.Interface, resyncPeriod time.Duration) cache.Sh
        tweakListOptions := func(options *metav1.ListOptions) {
                options.FieldSelector = selector
        }
-       return coreinformers.NewFilteredPodInformer(cs, metav1.NamespaceAll, resyncPeriod, nil, tweakListOptions)
+       return coreinformers.NewFilteredPodInformer(cs, metav1.NamespaceAll, resyncPeriod, cache.Indexers{}, tweakListOptions)
 }
```

### Kubernetes version

<details>

```console
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.2", GitCommit:"59603c6e503c87169aea6106f57b9f242f64df89", GitTreeState:"clean", BuildDate:"2020-01-18T23:30:10Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"23", GitVersion:"v1.23.5+3afdacb", GitCommit:"3c28e7a79b58e78b4c1dc1ab7e5f6c6c2d3aedd3", GitTreeState:"clean", BuildDate:"2022-05-10T16:30:48Z", GoVersion:"go1.17.5", Compiler:"gc", Platform:"linux/amd64"}
</details>

but applies also on latest master (see above)

### Cloud provider

<details>
any
</details>


### OS version

<details>
any
</details>


### Install tools

<details>
not relevant
</details>


### Container runtime (CRI) and version (if applicable)

<details>

</details>


### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

--- a/pkg/scheduler/scheduler.go
+++ b/pkg/scheduler/scheduler.go
@@ -452,10 +452,11 @@ func unionedGVKs(m map[framework.ClusterEvent]sets.String) map[framework.GVK]fra
 }
 
 // newPodInformer creates a shared index informer that returns only non-terminal pods.
+// The PodInformer allows indexers to be added, but note that only non-conflict indexers are allowed.
 func newPodInformer(cs clientset.Interface, resyncPeriod time.Duration) cache.SharedIndexInformer {
 	selector := fmt.Sprintf("status.phase!=%v,status.phase!=%v", v1.PodSucceeded, v1.PodFailed)
 	tweakListOptions := func(options *metav1.ListOptions) {
 		options.FieldSelector = selector
 	}
-	return coreinformers.NewFilteredPodInformer(cs, metav1.NamespaceAll, resyncPeriod, nil, tweakListOptions)
+	return coreinformers.NewFilteredPodInformer(cs, metav1.NamespaceAll, resyncPeriod, cache.Indexers{}, tweakListOptions)
 }

--- a/pkg/scheduler/scheduler_test.go
+++ b/pkg/scheduler/scheduler_test.go
@@ -28,6 +28,7 @@ import (
 	v1 "k8s.io/api/core/v1"
 	apierrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/client-go/informers"
 	"k8s.io/client-go/kubernetes"
@@ -471,4 +472,123 @@ func initScheduler(stop <-chan struct{}, cache internalcache.Cache, queue intern
 	)
 
 	return s, fwk, nil
+}
+
+func TestInitPluginsWithIndexers(t *testing.T) {
+	tests := []struct {
+		name string
+		// the plugin registration ordering must not matter, being map traversal random
+		entrypoints map[string]frameworkruntime.PluginFactory
+		wantErr     string
+	}{
+		{
+			name: "register indexer, no conflicts",
+			entrypoints: map[string]frameworkruntime.PluginFactory{
+				"AddIndexer": func(obj runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+					podInformer := handle.SharedInformerFactory().Core().V1().Pods()
+					err := podInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
+						"nodeName": indexByPodSpecNodeName,
+					})
+					return &TestPlugin{name: "AddIndexer"}, err
+				},
+			},
+		},
+		{
+			name: "register the same indexer name multiple times, conflict",
+			// order of registration doesn't matter
+			entrypoints: map[string]frameworkruntime.PluginFactory{
+				"AddIndexer1": func(obj runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+					podInformer := handle.SharedInformerFactory().Core().V1().Pods()
+					err := podInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
+						"nodeName": indexByPodSpecNodeName,
+					})
+					return &TestPlugin{name: "AddIndexer1"}, err
+				},
+				"AddIndexer2": func(obj runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+					podInformer := handle.SharedInformerFactory().Core().V1().Pods()
+					err := podInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
+						"nodeName": indexByPodAnnotationNodeName,
+					})
+					return &TestPlugin{name: "AddIndexer1"}, err
+				},
+			},
+			wantErr: "indexer conflict",
+		},
+		{
+			name: "register the same indexer body with different names, no conflicts",
+			// order of registration doesn't matter
+			entrypoints: map[string]frameworkruntime.PluginFactory{
+				"AddIndexer1": func(obj runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+					podInformer := handle.SharedInformerFactory().Core().V1().Pods()
+					err := podInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
+						"nodeName1": indexByPodSpecNodeName,
+					})
+					return &TestPlugin{name: "AddIndexer1"}, err
+				},
+				"AddIndexer2": func(obj runtime.Object, handle framework.Handle) (framework.Plugin, error) {
+					podInformer := handle.SharedInformerFactory().Core().V1().Pods()
+					err := podInformer.Informer().GetIndexer().AddIndexers(cache.Indexers{
+						"nodeName2": indexByPodAnnotationNodeName,
+					})
+					return &TestPlugin{name: "AddIndexer2"}, err
+				},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			fakeInformerFactory := NewInformerFactory(&fake.Clientset{}, 0*time.Second)
+
+			var registerPluginFuncs []st.RegisterPluginFunc
+			for name, entrypoint := range tt.entrypoints {
+				registerPluginFuncs = append(registerPluginFuncs,
+					// anything supported by TestPlugin is fine
+					st.RegisterFilterPlugin(name, entrypoint),
+				)
+			}
+			// we always need this
+			registerPluginFuncs = append(registerPluginFuncs,
+				st.RegisterQueueSortPlugin(queuesort.Name, queuesort.New),
+				st.RegisterBindPlugin(defaultbinder.Name, defaultbinder.New),
+			)
+			_, err := st.NewFramework(registerPluginFuncs, "test", frameworkruntime.WithInformerFactory(fakeInformerFactory))
+
+			if len(tt.wantErr) > 0 {
+				if err == nil || !strings.Contains(err.Error(), tt.wantErr) {
+					t.Errorf("got error %q, want %q", err, tt.wantErr)
+				}
+				return
+			}
+			if err != nil {
+				t.Fatalf("Failed to create scheduler: %v", err)
+			}
+		})
+	}
+}
+
+func indexByPodSpecNodeName(obj interface{}) ([]string, error) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		return []string{}, nil
+	}
+	if len(pod.Spec.NodeName) == 0 {
+		return []string{}, nil
+	}
+	return []string{pod.Spec.NodeName}, nil
+}
+
+func indexByPodAnnotationNodeName(obj interface{}) ([]string, error) {
+	pod, ok := obj.(*v1.Pod)
+	if !ok {
+		return []string{}, nil
+	}
+	if len(pod.Annotations) == 0 {
+		return []string{}, nil
+	}
+	nodeName, ok := pod.Annotations["node-name"]
+	if !ok {
+		return []string{}, nil
+	}
+	return []string{nodeName}, nil
 }

